pkg/sampling/oteltracestate.go

```diff
@@ -0,0 +1,225 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package sampling // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/sampling"
+
+import (
+	"fmt"
+	"io"
+	"regexp"
+	"strconv"
+)
+
+// OTelTraceState represents the `ot` section of the W3C tracestate
+// which is specified generically in https://opentelemetry.io/docs/specs/otel/trace/tracestate-handling/.
+type OTelTraceState struct {
+	commonTraceState
+
+	// sampling r and t-values
+	rnd       Randomness // r value parsed, as unsigned
+	rvalue    string     // 14 ASCII hex digits
+	threshold Threshold  // t value parsed, as a threshold
+	tvalue    string     // 1-14 ASCII hex digits
+}
+
+const (
+	// RName is the OTel tracestate field for R-value
+	RName = "rv"
+	// TName is the OTel tracestate field for T-value
+	TName = "th"
+
+	// hardMaxOTelLength is the maximum encoded size of an OTel
+	// tracestate value.
+	hardMaxOTelLength = 256
+
+	// chr        = ucalpha / lcalpha / DIGIT / "." / "_" / "-"
+	// ucalpha    = %x41-5A ; A-Z
+	// lcalpha    = %x61-7A ; a-z
+	// key        = lcalpha *(lcalpha / DIGIT )
+	// value      = *(chr)
+	// list-member = key ":" value
+	// list        = list-member *( ";" list-member )
+	otelKeyRegexp             = lcAlphaRegexp + lcDigitRegexp + `*`
+	otelValueRegexp           = `[a-zA-Z0-9._\-]*`
+	otelMemberRegexp          = `(?:` + otelKeyRegexp + `:` + otelValueRegexp + `)`
+	otelSemicolonMemberRegexp = `(?:` + `;` + otelMemberRegexp + `)`
+	otelTracestateRegexp      = `^` + otelMemberRegexp + otelSemicolonMemberRegexp + `*$`
+)
+
+var (
+	otelTracestateRe = regexp.MustCompile(otelTracestateRegexp)
+
+	otelSyntax = keyValueScanner{
+		maxItems:  -1,
+		trim:      false,
+		separator: ';',
+		equality:  ':',
+	}
+
+	// ErrInconsistentSampling is returned when a sampler update
+	// is illogical.  It is safe to ignore.  Samplers should avoid
```

Comment:
if it's safe to ignore why raise the error?

Reply:
I've updated the comment:
```
	// ErrInconsistentSampling is returned when a sampler update
	// is illogical, indicating that the tracestate was not
	// modified.  Preferrably Samplers will avoid seeing this
	// error by using a ThresholdGreater() test, which allows them
	// to report a more clear error to the user.  For example, if
	// data arrives sampled at 1/100 and an eqalizing sampler is
	// configureed for 1/2 sampling, the Sampler may detect the
	// illogical condition itself using ThresholdGreater and skip
	// the call to UpdateTValueWithSampling, which will have no
	// effect and return this error.  How a sampler decides to
	// handle this condition is up to the sampler: for example the
	// equalizing sampler can decide to pass through a span
	// indicating 1/100 sampling or it can reject the span.
```